### PR TITLE
Fix Issue 15768 - std.stdio.trustedStdout accesses __gshared data wit…

### PR DESCRIPTION
**Pull request by JackStouffer**
Running the following code

```d
import std.algorithm;
import std.parallelism;
import std.range;
import std.stdio;

void main()
{
    stderr = File("/dev/null", "w");
    
    foreach(t; 1_000_000.iota.parallel)
    {
        stderr.write("aaa");
    }
}
```

dies 100% of the time with v2.079.0, but with this fix it works every time I've tried.

I don't think there can be a test in the test suite for this, so please review carefully.

### CODE REVIEW

File: std/stdio.d [hunk 1]
```diff
@@ -843,12 +844,12 @@ Throws: $(D ErrnoException) on failure if closing the file.
     void detach() @safe
     {
         if (!_p) return;
-        if (_p.refs == 1)
+        if (atomicLoad(_p.refs) == 1)
```
Comment (n8sh) [inline std/stdio.d]: This is a potential resource leak. Example:

Two threads have references. Both check count, and see that it is 2. Both skip this branch and so the file is not closed. What needs to instead happen is unconditional `atomic!"-="` and check the result of that.

So in the new example:
Two threads have references. Both atomically decrement. One sees `0` as the new value and the other sees `1` as the new value. Whichever sees `0` then calls close.

File: std/stdio.d [hunk 2]
```diff
@@ -513,14 +514,16 @@ Throws: $(D ErrnoException) in case of error.
...
-        else if (_p.refs == 1)
         {
-            closeHandles();
-        }
-        else
-        {
-            _p.refs--;
-            _p = cast(Impl*) enforce(malloc(Impl.sizeof), "Out of memory");
+            atomicOp!"-="(_p.refs, 1);
+            if (atomicLoad(_p.refs) == 0)
```
Comment (n8sh) [inline std/stdio.d]: You can save an atomic operation: `if (atomicOp!"-="(_p.refs, 1) == 0)`

File: std/stdio.d [hunk 3]
```diff
@@ -887,8 +881,7 @@ Throws: $(D ErrnoException) on error.
         if (!_p) return; // succeed vacuously
         scope(exit)
         {
-            assert(_p.refs);
-            if (!--_p.refs)
+            if (atomicLoad(_p.refs) == 0)
```
Comment (andralex) [inline std/stdio.d]: This is different from the existing code - it doesn't decrement anymore.

**Comment by jercaianu** [comment]
@JackStouffer 
For testing, can't we redirect stdout to a file, write something in there from a couple of threads and then restore stdout?

**Comment by wilzbach** [comment]
> For testing, can't we redirect stdout to a file, write something in there from a couple of threads and then restore stdout?

Well, the test that Jack posted above allows to perfectly reproduce this bug. It just takes ~0.3s on my system. Maybe we can have a special folder (e.g. `stresstests`), s.t. these tests aren't run by default for the `unittest` target?

**Comment by JackStouffer** [comment]
Added test under a new version branch.

File: std/stdio.d [hunk 4]
```diff
@@ -3427,6 +3425,24 @@ void main()
...
+    // issue 15768
+    @system unittest
+    {
+        import std.parallelism : parallel;
+        import std.range : iota;
+
+        auto deleteme = testFilename();
+        stderr = File(deleteme, "w");
+
+        foreach(t; 1_000_000.iota.parallel)
```
Comment (wilzbach) [inline std/stdio.d]: std/stdio.d:3439:        foreach(t; 1_000_000.iota.parallel)
posix.mak:544: recipe for target 'style_lint' failed

**Comment by JackStouffer** [comment]
Can anyone that uses Windows confirm or deny that 13727 is fixed by this?

**Comment by JackStouffer** [comment]
I'd need to have the test be run on windows on the auto tester. And I'd assume we don't want to run that for every PR.

**Comment by wilzbach** [comment]
Well, we can definitely run it once here for validation?
(also the test case that Walter wanted to add it in dlang/dmd#5747 only uses 1000 runs.)

**Comment by anon17** [comment]
13727 has a C test, it doesn't depend on phobos.

**Comment by anon17** [comment]
Also don't remove comments about unsafe reassignment. You only fixed that reading them is thread-safe after single publishing, but concurrent writing is still subject to race condition.

**Comment by JackStouffer** [comment]
How so? Writing is done with LockingTextWriter, which uses OS file locks.

**Comment by schveiguy** [comment]
I think he means writing the handle itself, not writing data to the handle.

But the warning needs to be reworded -- the handle is subject to race conditions after it's initialized, for reading OR writing. It has nothing to do with the issue this PR is fixing.

**Comment by JackStouffer** [comment]
What specific race conditions are you talking about?

**Comment by schveiguy** [comment]
Let me back up a second here -- this PR is fixing an issue, but it doesn't appear to be the issue identified by the bug report's description. It appears that the test case given in the bug report has nothing to do with a race condition on modifying `stdout`, but improperly implemented reference counting inside `File`.

My recommendation: split the bug into another report, as there are really 2 bugs here -- one is the `File` refcount implementation, and the other is the race condition you could have from `trustedStdout` returning `stdout` without locking.

I'm not sure we even want to add a mutex lock to stdout. Most of the time, it doesn't change ever, or if it does, you are doing it early on when only one thread exists. Maybe just updating the documentation to reflect the rules is the proper fix.

**Comment by schveiguy** [comment]
> What specific race conditions are you talking about?

Classic race condition with shared data. 2 or more threads writing the `stdout` handle without locking, or 1 thread writing the `stdout` handle while another thread is calling `trustedStdout` to read it without locking. In order to properly provide a way to modify the `stdout` handle, we would need locking.

The other option is to document that you can only change the handles from a single thread, or you get race/undefined behavior.

**Comment by JackStouffer** [comment]
The test case shows the assignment outside of the threaded code because the bug in the bug report is specifically about the ref counting in File not working in a multi-threaded environment, and how that's an issue for stdout/in/err. It could possibly be an issue that a thread could re-assign stdout to another file instance in multi-threaded code, if that's what you're referring to.

In that case, the nature of stdout/in/err being global due to the nature of C libraries should be documented, and the fact that calls to write/ln auto lock the file. I really wouldn't call writing to stdout a "race condition" in the normal usage of the phrase because there are locks around writing, and just like in any well designed multi-threaded program you'll need some queue equivalent to make sure data is written to the file in the correct order. Once a chunk of data starts to be written to the file, that entire chunk will successfully write.

**Comment by schveiguy** [comment]
> It could possibly be an issue that a thread could re-assign stdout to another file instance in multi-threaded code, if that's what you're referring to.

Yes.

> I really wouldn't call writing to stdout a "race condition" in the normal usage of the phrase because there are locks around writing

Again, I'm not talking about writing data to stdout, but reassigning stdout itself to a different handle (e.g. `stdout = File("output.txt");`). There are no protections around this.

The original warning was worded poorly anyway -- it said reassigning stdout was dangerous due to bug 15768, but really any use of any of the standard handles (regardless of assignment) was dangerous.

In any case, your new information is good, so I think it's solved.

**Comment by schveiguy** [comment]
The title of the bug report is the problem (it's talking about `trustedStdout` which simply returns stdout). I'll just change it, so it doesn't confuse people.

**Comment by anon17** [comment]
`$(REF stdin,core,stdout)` you probably meant `$(REF stdin,core,stdio)`. Same for stderr.

**Pull request merged**